### PR TITLE
[GPU] Support buffer fusing for reshape and concat

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *buffer fusing for un-optimized reshape and concat*
 - *buffer fusing for optimized reshape and concat*
 - *Add unit tests for concat + reshape buffer fusing*
 - *Not allow buffer fusing when input_layout is the input of optimized reshape which is the parent's of concat in dynamic shape model*

### Tickets:
 - *118529*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp [hunk 1]
```diff
@@ -249,6 +282,24 @@ void concat_in_place_optimization::optimize_cascade(concatenation_node& node, st
...
+            auto reshape_node = dep.first;
+            bool is_dynamic = reshape_node->is_dynamic();
+            bool is_planar = format::is_default_format(reshape_node->get_output_layout().format);
+            bool no_pad = !reshape_node->get_output_layout().data_padding && !reshape_node->get_input_layouts().empty()
+                                && !reshape_node->get_input_layout(0).data_padding;
+            // if reshape is the input of concat and it is optimized out,
+            // update padding of dependency node of reshape
+            if (is_dynamic && is_planar && no_pad && !reshape_node->is_output() && !reshape_node->has_fused_primitives()) {
+                // reshape node is set to optimized out.
+                reshape_node->can_be_optimized(true);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp:295]: 1. What is the policy? 
a. support [optimized reshape] + concat only. 
- a.1 Here target reshape is opt out in the previous pass. 
- a.2. Even though reshape is not marked as optimized out from the previous pass, mark it as optimize here. 
b. support both [optimized reshape] + concat and [not optimized reshape] + concat
- b.1 same as a1
- b2 same as a2

File: src/plugins/intel_gpu/src/graph/network.cpp [hunk 2]
```diff
@@ -836,6 +840,10 @@ void network::build_exec_order() {
                     // i.e., We need to make sure that all the preds of them are already updated too.
                     for (auto dep : node->get_dependencies()) {
                         if (!dep.first->is_type<data>()) {
+                            if (dep.first->is_type<reshape>() && dep.first->can_be_optimized()) {
+                                auto reshape_node = dep.first;
+                                add_to_exec_order(reshape_node->get_dependency(0).id());
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/network.cpp:845]: Not adding reshape itself?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/network.cpp:845]: It's not reshape itself, it is the input of reshape node.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_buffer_fusing.cpp]: Fixed it by offline comments. thanks.

**Comment by yeonbok** [comment]
LGTM overall
1. Please resolve unittest failures 
2. Please check network accuracy

**Pull request closed without merging**